core/lib/pattern_exporter.js

```diff
@@ -14,18 +15,50 @@ var pattern_exporter = function () {
   function exportPatterns(patternlab) {
     //read the config export options
     var exportPartials = patternlab.config.patternExportPatternPartials;
+    var exportAll = patternlab.config.patternExportAll;
+
+    if (exportAll) {
+      for (var i = 0; i < patternlab.patterns.length; i++) {
+        if (!patternlab.patterns[i].patternPartial.startsWith('-')) {
+          exportSinglePattern(patternlab, patternlab.patterns[i]);
+        }
+      }
+
+      return;
+    }
 
     //find the chosen patterns to export
     for (var i = 0; i < exportPartials.length; i++) {
       for (var j = 0; j < patternlab.patterns.length; j++) {
         if (exportPartials[i] === patternlab.patterns[j].patternPartial) {
           //write matches to the desired location
-          fs.outputFileSync(patternlab.config.patternExportDirectory + patternlab.patterns[j].patternPartial + '.html', patternlab.patterns[j].patternPartialCode);
+          exportSinglePattern(patternlab, patternlab.patterns[j]);
         }
       }
     }
   }
 
+  function exportSinglePattern(patternlab, pattern) {
+    var preserveDirStructure = patternlab.config.patternExportPreserveDirectoryStructure;
+    var patternName = pattern.patternPartial;
+    var patternDir = patternlab.config.patternExportDirectory;
+    var patternCode = pattern.patternPartialCode;
+    var patternFileExtension = ".html";
+    if (preserveDirStructure) {
+      // Extract the first part of the pattern partial as the directory in which
+      // it should go.
+      patternDir = path.join(patternDir, pattern.patternPartial.split('-')[0]);
+      patternName = pattern.patternPartial.split('-').slice(1).join('-');
+    }
+
+    if (patternlab.config.patternExportRaw) {
+      patternCode = pattern.extendedTemplate;
+      patternFileExtension = "." + JSON.parse(pattern.patternData).patternExtension;
```

Comment:
I have no idea what that JSON.parse is doing in there.
 
`pattern.patternExtension` suffices.

Reply:
There is no ```pattern.patternExtension``` - "patternExtension" only lives within ```pattern.patternData``` - which does seem lto be a stringified JSON object (hence the parse @jwir3 added).

We do have a ```pattern.fileExtension``` available (which shows up with the period in it, such as ```.hbs```. I'm guessing we could use this to clean it up, but I'm not sure to why we have this difference between fileExtension and patternExtension, and why patternExtension is only available in that weird string JSON patternData.